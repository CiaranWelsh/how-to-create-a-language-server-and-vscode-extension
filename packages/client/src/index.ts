import {
  LanguageClient,
  LanguageClientOptions,
  ServerOptions
} from 'vscode-languageclient';

let client: LanguageClient;

export function activate() {
  const serverModule = require.resolve(
    '@donaldpipowitch/vscode-extension-server'
  );

  // Debug options for server are used when we launch the extension in debug mode
  const serverOptions: ServerOptions = {
    run: { module: serverModule },
    debug: {
      module: serverModule,
      options: { execArgv: ['--inspect=6009'] }
    }
  };

  const clientOptions: LanguageClientOptions = {
    documentSelector: [
      {
        pattern: '**/.vscode/extensions.json',
        scheme: 'file'
      }
    ]
  };

  client = new LanguageClient(
    'vscode-extensions-files',
    'VS Code Extension Client',
    serverOptions,
    clientOptions
  );

  // Starting the client will also launch the server.
  client.start();
}

export function deactivate() {
  if (client) {
    return client.stop();
  }
}
