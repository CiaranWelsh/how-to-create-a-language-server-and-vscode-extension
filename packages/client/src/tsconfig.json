{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": {
    "declarationDir": "../dist",
    "outDir": "../dist",
    "types": ["node"]
  },
  "references": [{ "path": "../../server/src" }]
}
