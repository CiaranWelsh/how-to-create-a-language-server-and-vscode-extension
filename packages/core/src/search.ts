import axios, { Canceler } from 'axios';

export { Canceler };

/**
 * In the `.vscode/extensions.json` we'll need to use `${publisher.publisherName}.${extensionName}`.
 *
 * @example
 * {
 *   "publisher": {
 *     "publisherId": "d16f4e39-2ffb-44e3-9c0d-79d873570e3a",
 *     "publisherName": "esbenp",
 *     "displayName": "Esben Petersen",
 *     "flags": "none"
 *   },
 *   "extensionId": "96fa4707-6983-4489-b7c5-d5ffdfdcce90",
 *   "extensionName": "prettier-vscode",
 *   "displayName": "Prettier - Code formatter",
 *   "flags": "validated, public",
 *   "lastUpdated": "2018-08-09T12:05:04.413Z",
 *   "publishedDate": "2017-01-10T19:52:02.703Z",
 *   "releaseDate": "2017-01-10T19:52:02.703Z",
 *   "shortDescription": "VS Code plugin for prettier/prettier",
 *   "deploymentType": 0
 * }
 */
export type Extension = {
  publisher: {
    publisherId: string;
    publisherName: string;
    displayName: string;
    flags: string;
  };
  extensionId: string;
  extensionName: string;
  displayName: string;
  flags: string;
  lastUpdated: string;
  publishedDate: string;
  releaseDate: string;
  shortDescription: string;
  deploymentType: number;
};

export type SearchRequest = {
  cancel: Canceler;
  request: Promise<Extension[] | void>;
};

export function search(value: string): SearchRequest {
  const { token, cancel } = axios.CancelToken.source();
  const options = {
    cancelToken: token,
    url:
      'https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery',
    method: 'post',
    headers: {
      accept: 'application/json;api-version=5.0-preview.1;excludeUrls=true'
    },
    data: {
      filters: [
        {
          criteria: [
            // which visual studio app? code
            { filterType: 8, value: 'Microsoft.VisualStudio.Code' },
            // our search value
            { filterType: 10, value }
          ],
          pageSize: 10,
          pageNumber: 1
        }
      ]
    }
  };
  const request = axios(options)
    .then(({ data }) => data.results[0].extensions as Extension[])
    .catch((error) => {
      if (!axios.isCancel(error)) {
        throw error;
      }
    });
  return { request, cancel };
}
