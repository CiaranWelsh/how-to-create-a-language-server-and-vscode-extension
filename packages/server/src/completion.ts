import { search, Canceler } from '@donaldpipowitch/vscode-extension-core';
import {
  TextDocuments,
  CompletionItemKind,
  Connection
} from 'vscode-languageserver';
import { parseTree, findNodeAtOffset, Node } from 'jsonc-parser';

let lastCancel: Canceler | null = null;

export function configureCompletion(
  connection: Connection,
  documents: TextDocuments
) {
  // The onCompletion handler provides the initial list of the completion items.
  // This is an example of the params passed to the handler:
  // {
  //   "textDocument": {
  //     "uri": "file:///Users/pipo/workspace/test-txt/test.txt"
  //   },
  //   "position": {
  //     "line": 0,
  //     "character": 10
  //   },
  //   "context": {
  //     "triggerKind": 1
  //   }
  // }
  connection.onCompletion(async ({ position, textDocument }) => {
    if (lastCancel) {
      lastCancel();
      lastCancel = null;
    }

    if (!textDocument.uri.endsWith('.vscode/extensions.json')) return;

    const document = documents.get(textDocument.uri);
    if (!document) return;

    const tree = parseTree(document.getText());
    const node = findNodeAtOffset(tree, document.offsetAt(position));

    if (!isExtensionValue(node)) return;

    // only search for queries with at least three characters
    if (node.value && node.value.length <= 2) return;

    const query = node.value;
    // connection.console.log(`You searched for ${query}.`);

    const { cancel, request } = search(query);
    lastCancel = cancel;

    const extensions = await request;
    lastCancel = null;

    if (!extensions) return;

    const completionItems = extensions.map(
      ({
        publisher: { publisherName },
        extensionName,
        displayName,
        shortDescription
      }) => ({
        // `"` hack, because JSON treats `"` as part of the completion here
        // see https://github.com/Microsoft/vscode-extension-samples/issues/93#issuecomment-429849514
        label: `"${publisherName}.${extensionName}`,
        kind: CompletionItemKind.Text,
        detail: displayName,
        documentation: shortDescription
      })
    );

    return completionItems;
  });
}

function isExtensionValue(node: Node | undefined): node is Node {
  // no node
  if (!node) return false;
  // not a string node
  if (node.type !== 'string') return false;
  // not within an array
  if (!node.parent) return false;
  if (node.parent.type !== 'array') return false;
  // not on a "recommendations" or "unwantedRecommendations" property
  if (!node.parent.parent) return false;
  if (node.parent.parent.type !== 'property') return false;
  if (
    node.parent.parent.children![0].value !== 'recommendations' &&
    node.parent.parent.children![0].value !== 'unwantedRecommendations'
  )
    return false;
  // not on an object
  if (!node.parent.parent.parent) return false;
  if (node.parent.parent.parent.type !== 'object') return false;
  // not the root
  if (node.parent.parent.parent.parent) return false;
  // whoohoo
  return true;
}
