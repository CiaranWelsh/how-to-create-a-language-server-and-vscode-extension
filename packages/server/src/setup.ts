import {
  createConnection,
  TextDocuments,
  ProposedFeatures
} from 'vscode-languageserver';

const connection = createConnection(ProposedFeatures.all);

// text document manager (supports full document sync only)
const documents = new TextDocuments();

connection.onInitialize(() => ({
  // tells the client what we support:
  // - full text sync
  // - code completion
  capabilities: {
    textDocumentSync: documents.syncKind,
    completionProvider: {
      resolveProvider: true
    }
  }
}));

// Listen for events
documents.listen(connection);
connection.listen();

export { connection, documents };
