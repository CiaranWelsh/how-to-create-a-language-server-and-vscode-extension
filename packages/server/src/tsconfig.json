{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": {
    "declarationDir": "../dist",
    "outDir": "../dist",
    "types": ["vscode", "node"]
  },
  "references": [{ "path": "../../core/src" }]
}
