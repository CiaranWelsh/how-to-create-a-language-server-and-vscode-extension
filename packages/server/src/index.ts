import { connection, documents } from './setup';
import { configureCompletion } from './completion';

configureCompletion(connection, documents);
