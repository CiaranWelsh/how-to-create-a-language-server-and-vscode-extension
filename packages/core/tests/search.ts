import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { search } from '../src/search';
import prettierResponse from './__mocks__/prettier-response.json';

const mock = new MockAdapter(axios);

test('should search extensions', async () => {
  mock.onAny().replyOnce(200, prettierResponse);
  expect(await search('prettier').request).toMatchSnapshot();
});

test('should cancel search', async () => {
  mock.onAny().replyOnce(200, prettierResponse);
  const { request, cancel } = search('prettier');
  cancel();
  expect(await request).toBe(undefined);
});
