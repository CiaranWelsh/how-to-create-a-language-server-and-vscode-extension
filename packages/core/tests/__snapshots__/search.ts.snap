// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should search extensions 1`] = `
Array [
  Object {
    "deploymentType": 0,
    "displayName": "Prettier - Code formatter",
    "extensionId": "96fa4707-6983-4489-b7c5-d5ffdfdcce90",
    "extensionName": "prettier-vscode",
    "flags": "validated, public",
    "lastUpdated": "2018-08-09T12:05:04.413Z",
    "publishedDate": "2017-01-10T19:52:02.703Z",
    "publisher": Object {
      "displayName": "Esben Petersen",
      "flags": "none",
      "publisherId": "d16f4e39-2ffb-44e3-9c0d-79d873570e3a",
      "publisherName": "esbenp",
    },
    "releaseDate": "2017-01-10T19:52:02.703Z",
    "shortDescription": "VS Code plugin for prettier/prettier",
  },
  Object {
    "deploymentType": 0,
    "displayName": "Prettier Now",
    "extensionId": "707c46ed-f0c5-4fba-9f7f-025257b23b6b",
    "extensionName": "prettier-now",
    "flags": "validated, public",
    "lastUpdated": "2017-09-19T07:55:32.053Z",
    "publishedDate": "2017-06-08T07:55:35.41Z",
    "publisher": Object {
      "displayName": "Remi Marsal",
      "flags": "none",
      "publisherId": "af4a4a8d-7c71-4bdd-aea6-06b83fb8e176",
      "publisherName": "remimarsal",
    },
    "releaseDate": "2017-06-08T07:55:35.41Z",
    "shortDescription": "VS Code plugin for Prettier Miscellaneous / Code Formatter",
  },
  Object {
    "deploymentType": 0,
    "displayName": "Prettier - JavaScript formatter",
    "extensionId": "7a6db49e-50ff-4e96-b35d-3042129cc632",
    "extensionName": "prettier-vscode-with-tabs",
    "flags": "validated, public",
    "lastUpdated": "2017-03-13T14:29:40.73Z",
    "publishedDate": "2017-02-20T13:45:33.267Z",
    "publisher": Object {
      "displayName": "Bastian Kistner",
      "flags": "none",
      "publisherId": "81e36226-0c30-4890-a345-2bce2fdc2278",
      "publisherName": "passionkind",
    },
    "releaseDate": "2017-02-20T13:45:33.267Z",
    "shortDescription": "VS Code plugin for jlongster/prettier with tabs support",
  },
  Object {
    "deploymentType": 0,
    "displayName": "Prettier - JavaScript formatter",
    "extensionId": "db0e58d0-3519-4d7f-b181-11690e64588a",
    "extensionName": "prettier-vscode-space-parenthesis",
    "flags": "validated, public",
    "lastUpdated": "2017-05-11T12:34:44.88Z",
    "publishedDate": "2017-05-04T15:06:53.317Z",
    "publisher": Object {
      "displayName": "Mathieu SCHROETER",
      "flags": "none",
      "publisherId": "c71ad5a9-824a-4721-a6c4-db53381e6955",
      "publisherName": "schroeter",
    },
    "releaseDate": "2017-05-04T15:06:53.317Z",
    "shortDescription": "Fork of prettier-vscode: VS Code plugin for Skywalker13/prettier-space-parenthesis",
  },
  Object {
    "deploymentType": 0,
    "displayName": "Prettier - JavaScript formatter",
    "extensionId": "1bf1fc66-34d9-4ff6-98eb-bb723ae2c228",
    "extensionName": "prettier-vscode-standard",
    "flags": "validated, public",
    "lastUpdated": "2017-08-08T09:01:17.703Z",
    "publishedDate": "2017-07-08T20:03:57.16Z",
    "publisher": Object {
      "displayName": "bySabi Files",
      "flags": "none",
      "publisherId": "e6f41c7c-c4a5-47c1-af02-95a3ede31236",
      "publisherName": "bysabi",
    },
    "releaseDate": "2017-07-08T20:03:57.16Z",
    "shortDescription": "prettier or prettier + standard --fix",
  },
  Object {
    "deploymentType": 0,
    "displayName": "Prettier - JavaScript formatter",
    "extensionId": "18299cdc-11ff-4089-8b92-7a497eb14830",
    "extensionName": "prettier-vscode-semistandard",
    "flags": "validated, public",
    "lastUpdated": "2017-08-08T09:03:23.713Z",
    "publishedDate": "2017-07-08T19:04:10.22Z",
    "publisher": Object {
      "displayName": "bySabi Files",
      "flags": "none",
      "publisherId": "e6f41c7c-c4a5-47c1-af02-95a3ede31236",
      "publisherName": "bysabi",
    },
    "releaseDate": "2017-07-08T19:04:10.22Z",
    "shortDescription": "prettier or prettier + semistandard --fix",
  },
  Object {
    "deploymentType": 0,
    "displayName": "Prettier Fuge Standard",
    "extensionId": "cebdce3f-66e8-4a37-8793-9bd993e55514",
    "extensionName": "prettier-fuge-standard",
    "flags": "validated, public",
    "lastUpdated": "2018-02-25T04:55:56.447Z",
    "publishedDate": "2018-02-23T06:37:41.917Z",
    "publisher": Object {
      "displayName": "clmystes",
      "flags": "none",
      "publisherId": "bcc497b3-7e67-4b58-92a8-5e5c47024fc5",
      "publisherName": "clmystes",
    },
    "releaseDate": "2018-02-23T06:37:41.917Z",
    "shortDescription": "prettier with fuge-standard",
  },
  Object {
    "deploymentType": 0,
    "displayName": "Prettier-Standard - JavaScript formatter",
    "extensionId": "945fe310-535c-4f65-b93d-b93d70401ab8",
    "extensionName": "prettier-standard-vscode",
    "flags": "validated, public",
    "lastUpdated": "2017-04-20T19:56:57.293Z",
    "publishedDate": "2017-03-13T16:43:47.693Z",
    "publisher": Object {
      "displayName": "numso",
      "flags": "none",
      "publisherId": "47645c15-4969-4123-9fe0-78953af30f9d",
      "publisherName": "numso",
    },
    "releaseDate": "2017-03-13T16:43:47.693Z",
    "shortDescription": "VS Code plugin for prettier + standard",
  },
  Object {
    "deploymentType": 0,
    "displayName": "Prettier-Standard - JavaScript formatter",
    "extensionId": "b945bf5e-a118-43c4-97a4-4a9b9c8b1eb8",
    "extensionName": "vscode-prettier-standard",
    "flags": "validated, public",
    "lastUpdated": "2017-11-27T16:28:23.42Z",
    "publishedDate": "2017-11-27T15:42:03.343Z",
    "publisher": Object {
      "displayName": "iSayme",
      "flags": "none",
      "publisherId": "db59c0b4-649b-42c4-9e6c-8325baceec56",
      "publisherName": "iSayme",
    },
    "releaseDate": "2017-11-27T15:42:03.343Z",
    "shortDescription": "VSCode plugin for prettier standard",
  },
  Object {
    "deploymentType": 0,
    "displayName": "MaxLint",
    "extensionId": "9f8dd9e3-3ac8-4038-983d-822267df8518",
    "extensionName": "MaxLint",
    "flags": "validated, public",
    "lastUpdated": "2018-09-14T09:20:12.463Z",
    "publishedDate": "2018-08-29T10:43:06.793Z",
    "publisher": Object {
      "displayName": "Yusuf Aksoy",
      "flags": "none",
      "publisherId": "35832125-0ff1-43de-b3e6-6a3a978b8f6d",
      "publisherName": "yaksoy",
    },
    "releaseDate": "2018-08-29T10:43:06.793Z",
    "shortDescription": "Maximum automatic linting by running lebab, Prettier, and ESLint",
  },
]
`;
